### PR TITLE
REFACTOR-#2748: CSV IO part refactoring

### LINKED ISSUE #2748: CSV IO part refactoring
Currently CSV IO part [`_read`](https://github.com/modin-project/modin/blob/master/modin/engines/base/io/text/csv_dispatcher.py#L26) function is quite big function, that making it difficult to understand it logic, so this function can be spitted into functions to fix it.

### PR DESCRIPTION
**Pull request by amyskov**
## What do these changes do?

This PR refactores `csv_dispatcher.py` module, most significant changes are:
- Some code blocks moved into functions.
- Fallback-to-pandas checks grouped and moved on the top of `_read` function.
- Modification of original `read_csv` kwargs parameters was removed

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2748  
- [x] tests passing

### CODE REVIEW

**Comment by amyskov** [comment]
Measurements results are next:
```
   before           after      ratio (lower is better)
 [d1321f5f]       [4f2a1457]
 148±10ms          147±6ms     1.00  io.csv.TimeReadCsvSkiprows.time_skiprows((1000000, 10), 500)
  451±5ms          459±4ms     1.02  io.csv.TimeReadCsvSkiprows.time_skiprows((1000000, 10), lambda x: x % 2, default_to_pandas)
  141±5ms         151±10ms     1.07  io.csv.TimeReadCsvSkiprows.time_skiprows((1000000, 10), None)
2.21±0.03s       2.18±0.01s     0.98  io.csv.TimeReadCsvSkiprows.time_skiprows((5000, 5000), 500)
1.45±0.01s       1.44±0.01s     0.99  io.csv.TimeReadCsvSkiprows.time_skiprows((5000, 5000), lambda x: x % 2, default_to_pandas)
2.13±0.02s       2.12±0.03s     1.00  io.csv.TimeReadCsvSkiprows.time_skiprows((5000, 5000), None)
```
So, refactoring doesn't affected read_csv performance.

File: asv_bench/benchmarks/io/csv.py [hunk 1]
```diff
@@ -53,6 +53,14 @@ class TimeReadCsvSkiprows(BaseReadCsv):
...
         [
             None,
+            # integer value
+            min(
+                [
+                    UNARY_OP_DATA_SIZE[ASV_DATASET_SIZE][shape_id][0]
+                    for shape_id in range(len(UNARY_OP_DATA_SIZE[ASV_DATASET_SIZE]))
+                ]
+            )
+            // 10,
```
Comment (anmyachev) [inline asv_bench/benchmarks/io/csv.py:63]: Please make changes in ASV always in separate PR.

File: modin/engines/base/io/text/csv_dispatcher.py [hunk 2]
```diff
@@ -12,44 +12,29 @@
...
-import sys
+from csv import QUOTE_NONE
 
 from modin.config import NPartitions
 
 
 class CSVDispatcher(TextFileDispatcher):
     @classmethod
     def _read(cls, filepath_or_buffer, **kwargs):
+        # Fallback to pandas if parameter is not supported in Modin yet
```
Comment (anmyachev) [inline modin/engines/base/io/text/csv_dispatcher.py:24]: It's better to move check parameters support in separate function - for example: `_to_csv_check_support` in https://github.com/modin-project/modin/pull/2715

File: modin/engines/base/io/text/csv_dispatcher.py [hunk 3]
```diff
@@ -12,44 +12,29 @@
...
-                or compression_type == "bz2"
-                or compression_type == "xz"
-            ):
-                kwargs["compression"] = compression_type
-            elif (
-                compression_type == "zip"
-                and sys.version_info[0] == 3
-                and sys.version_info[1] >= 7
-            ):
-                # need python3.7 to .seek and .tell ZipExtFile
```
Comment (anmyachev) [inline modin/engines/base/io/text/csv_dispatcher.py:49]: Don't remove comment.

File: modin/engines/base/io/text/csv_dispatcher.py [hunk 4]
```diff
@@ -72,7 +80,7 @@ def _read(cls, filepath_or_buffer, **kwargs):
                 filepath_or_buffer,
                 **dict(kwargs, usecols=None, nrows=0, skipfooter=0, index_col=None),
             ).columns
-        elif index_col is None and not usecols:
+        elif index_col is None and not kwargs.get("usecols", None):
```
Comment (anmyachev) [inline modin/engines/base/io/text/csv_dispatcher.py:83]: You already define `usecols` variable, why not used here?

File: modin/engines/base/io/text/csv_dispatcher.py [hunk 5]
```diff
@@ -83,115 +91,141 @@ def _read(cls, filepath_or_buffer, **kwargs):
                 index_col = list(range(num_cols - len(names)))
                 if len(index_col) == 1:
                     index_col = index_col[0]
-                kwargs["index_col"] = index_col
```
Comment (anmyachev) [inline modin/engines/base/io/text/csv_dispatcher.py:86]: if you delete this line, then why create an empty dataframe on the previous lines?

File: modin/engines/base/io/text/csv_dispatcher.py [hunk 6]
```diff
@@ -83,115 +91,141 @@ def _read(cls, filepath_or_buffer, **kwargs):
...
-                    for i in range(num_splits)
-                ]
-
-            args = {
-                "fname": filepath_or_buffer,
-                "num_splits": num_splits,
-                **partition_kwargs,
-            }
+            rows_skipper = cls.rows_skipper_builder(f, quotechar, is_quoting=is_quoting)
+            rows_skipper(skiprows)
```
Comment (anmyachev) [inline modin/engines/base/io/text/csv_dispatcher.py:127]: Why not directly call `_read_rows`?

File: modin/engines/base/io/text/csv_dispatcher.py [hunk 7]
```diff
@@ -83,115 +91,141 @@ def _read(cls, filepath_or_buffer, **kwargs):
...
+            column_widths,
+            dtypes=dtypes,
+        )
+
+        new_query_compiler = cls._get_new_qc(
+            new_frame,
+            kwargs.get("skipfooter", None),
+            kwargs.get("squeeze", False),
+            index_col,
+        )
```
Comment (anmyachev) [inline modin/engines/base/io/text/csv_dispatcher.py:176]: All this can be moved into the function for creating query_compiler.

**Review by anmyachev — CHANGES_REQUESTED** [review]

File: asv_bench/benchmarks/io/csv.py (hunk 1 above)
Comment (amyskov) [reply asv_bench/benchmarks/io/csv.py]: Ok, ASV changes were removed.

File: modin/engines/base/io/text/csv_dispatcher.py (hunk 2 above)
Comment (amyskov) [reply modin/engines/base/io/text/csv_dispatcher.py]: Ok, function was added.

File: modin/engines/base/io/text/csv_dispatcher.py (hunk 3 above)
Comment (amyskov) [reply modin/engines/base/io/text/csv_dispatcher.py:49]: Comment was restored.

File: modin/engines/base/io/text/csv_dispatcher.py (hunk 4 above)
Comment (amyskov) [reply modin/engines/base/io/text/csv_dispatcher.py]: Right, modified `usecols` can be used here. Fixed.

File: modin/engines/base/io/text/csv_dispatcher.py (hunk 5 above)
Comment (amyskov) [reply modin/engines/base/io/text/csv_dispatcher.py:86]: `index_col` still will be in the `partitions_kwargs` (`index_col` added to `partiotion_kwargs` [here](https://github.com/amyskov/modin/blob/text_io_refactoring/modin/engines/base/io/text/csv_dispatcher.py#L112)). This line was removed to avoid changing original `kwargs`.

File: modin/engines/base/io/text/csv_dispatcher.py (hunk 6 above)
Comment (amyskov) [reply modin/engines/base/io/text/csv_dispatcher.py:117]: There is no special reason for this, is is just to make `_read` function lighter(in the future `rows_skipper` can be used several times).

File: modin/engines/base/io/text/csv_dispatcher.py (hunk 7 above)
Comment (amyskov) [reply modin/engines/base/io/text/csv_dispatcher.py]: Right, fixed.

File: modin/engines/base/io/text/csv_dispatcher.py [hunk 8]
```diff
@@ -83,115 +82,163 @@ def _read(cls, filepath_or_buffer, **kwargs):
...
+
+        with cls.file_open(filepath_or_buffer_md, "rb", compression_infered) as f:
+            rows_skipper = cls.rows_skipper_builder(f, quotechar, is_quoting=is_quoting)
+            rows_skipper(skiprows)
 
             splits = cls.partitioned_file(
                 f,
                 num_partitions=num_partitions,
-                nrows=nrows,
-                skiprows=skiprows,
```
Comment (anmyachev) [inline modin/engines/base/io/text/csv_dispatcher.py:166]: why not just skip rows inside `partitioned_file` func?

Comment (amyskov) [reply modin/engines/base/io/text/csv_dispatcher.py:166]: This is done to make `partitioned_file` function lighter. The plan is to handle trivial integer `skiprows` in the `CSVDispatcher._read` function and complicated `skiprows` cases (callables or arrays) in the `TextFileDispatcher.partitioned_file` function.

Comment (anmyachev) [reply modin/engines/base/io/text/csv_dispatcher.py:166]: Why solve one problem in two different places?

Comment (amyskov) [reply modin/engines/base/io/text/csv_dispatcher.py:166]: The reason is that when `skiprows` will be non-integer `partitioned_file` will become complicated, but, i think, you are right - handling of `skiprows` should be done in one place, will fix it soon.

**Review by anmyachev — DISMISSED** [review]

**Comment by amyskov** [comment]
Hi @devin-petersohn, can this PR be merged or you have any comments?

**Comment by amyskov** [comment]
Hi @devin-petersohn please take a look at this PR.

File: modin/engines/base/io/text/text_file_dispatcher.py [hunk 9]
```diff
@@ -258,3 +259,157 @@ def _read_rows(
...
+            num_splits = 1
+        else:
+            column_widths = [
+                column_chunksize
+                if len(column_names) > (column_chunksize * (i + 1))
+                else 0
+                if len(column_names) < (column_chunksize * i)
+                else len(column_names) - (column_chunksize * i)
+                for i in range(num_splits)
+            ]
```
Comment (williamma12) [inline modin/engines/base/io/text/text_file_dispatcher.py:379]: Would this be more readable if we used normal if statements rather than inline if statements?

File: modin/engines/base/io/text/csv_dispatcher.py [hunk 10]
```diff
@@ -83,115 +82,161 @@ def _read(cls, filepath_or_buffer, **kwargs):
...
+            squeeze=kwargs.get("squeeze", False),
+            skipfooter=kwargs.get("skipfooter", None),
+            parse_dates=kwargs.get("parse_dates", False),
+        )
+        return new_query_compiler
+
+    # _read helper functions
+    @classmethod
+    def _read_csv_check_support(
+        cls, filepath_or_buffer, read_csv_kwargs, compression_infered
```
Comment (williamma12) [inline modin/engines/base/io/text/csv_dispatcher.py:146]: Could we add type hinting for this or are the arguments too tedious to add type hints for (if this is the case, my bet would be that `read_csv_kwargs` might be challenging?)? 

If `read_csv_kwargs` is too hard to had type hints for, would it affect readability if we only passed in the arguments needed? I think we only need a couple arguments from `read_csv_kwargs`

File: modin/engines/base/io/text/text_file_dispatcher.py [hunk 11]
```diff
@@ -258,3 +259,157 @@ def _read_rows(
...
+        kwargs:
+                read_csv kwargs.
+
+        Returns
+        -------
+        usecols:
+                validated/converted usecols parameter.
+        """
+        usecols = kwargs.get("usecols", None)
+        usecols_md = _validate_usecols_arg(usecols)
```
Comment (williamma12) [inline modin/engines/base/io/text/text_file_dispatcher.py:329]: Would it help readability if we did something like this so that the results of [`_validate_usecols_arg`](https://github.com/pandas-dev/pandas/blob/d01561fb7b9ad337611fa38a3cfb7e8c2faec608/pandas/io/parsers/base_parser.py#L813)
would be clearer?

```suggestion
        verified_usecols, usecols_dtype = _validate_usecols_arg(usecols)
```

File: modin/engines/base/io/text/text_file_dispatcher.py [hunk 12]
```diff
@@ -258,3 +259,157 @@ def _read_rows(
...
+                    nrows=n,
+                )[1]
+
+        return skipper
+
+    @classmethod
+    def _define_header_size(
+        cls,
+        header: Union[int, Sequence[int], str, None] = "infer",
+        names: Union[Sequence, None] = None,
```
Comment (williamma12) [inline modin/engines/base/io/text/text_file_dispatcher.py:283]: Can we use `Optional` here as follows?
```suggestion
        names: Optional[Sequence] = None,
```

File: modin/engines/base/io/text/text_file_dispatcher.py [hunk 13]
```diff
@@ -258,3 +259,157 @@ def _read_rows(
...
+                    is_quoting=is_quoting,
+                    nrows=n,
+                )[1]
+
+        return skipper
+
+    @classmethod
+    def _define_header_size(
+        cls,
+        header: Union[int, Sequence[int], str, None] = "infer",
```
Comment (williamma12) [inline modin/engines/base/io/text/text_file_dispatcher.py:282]: I dont believe header can be `None`? Or is it [undocumented behavior](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html)?

File: modin/engines/base/io/text/csv_dispatcher.py [hunk 14]
```diff
@@ -210,11 +255,63 @@ def _read(cls, filepath_or_buffer, **kwargs):
...
+        cls,
+        partition_ids,
+        index_ids,
+        dtypes_ids,
+        index_col_md,
+        index_name,
+        column_widths,
+        column_names,
+        **kwargs,
+    ):
```
Comment (williamma12) [inline modin/engines/base/io/text/csv_dispatcher.py:272]: Would it be possible to add type hints here or does type hinting not play nice with `new_query_compiler`?

File: modin/engines/base/io/text/csv_dispatcher.py [hunk 15]
```diff
@@ -83,115 +82,161 @@ def _read(cls, filepath_or_buffer, **kwargs):
...
+            return False
+
+        skiprows = read_csv_kwargs.get("skiprows")
+        if skiprows is not None and not isinstance(skiprows, int):
+            return False
+
+        return True
+
+    @classmethod
+    def _define_index(cls, index_ids, index_col, index_name):
```
Comment (williamma12) [inline modin/engines/base/io/text/csv_dispatcher.py:190]: Would it be possible to add type hints here?

File: modin/engines/base/io/text/csv_dispatcher.py [hunk 16]
```diff
@@ -83,115 +82,161 @@ def _read(cls, filepath_or_buffer, **kwargs):
...
+        index_name:
+                Name that should be assigned to the index if `index_col`
+                is not provided.
+
+        Returns
+        -------
+        new_index:
+                Index that should be passed to the new_frame constructor.
+        row_lengths:
+                Partitions rows lenths.
```
Comment (williamma12) [inline modin/engines/base/io/text/csv_dispatcher.py:208]:
```suggestion
                Partitions rows lengths.
```

File: modin/engines/base/io/text/csv_dispatcher.py [hunk 17]
```diff
@@ -83,115 +82,161 @@ def _read(cls, filepath_or_buffer, **kwargs):
...
-        # Compute dtypes by getting collecting and combining all of the partitions. The
-        # reported dtypes from differing rows can be different based on the inference in
-        # the limited data seen by each worker. We use pandas to compute the exact dtype
-        # over the whole column for each column. The index is set below.
-        dtypes = cls.get_dtypes(dtypes_ids) if len(dtypes_ids) > 0 else None
+        return new_index, row_lengths
 
-        partition_ids = cls.build_partition(partition_ids, row_lengths, column_widths)
+    @classmethod
+    def _define_column_names(cls, column_names, parse_dates, column_widths):
```
Comment (williamma12) [inline modin/engines/base/io/text/csv_dispatcher.py:222]: Would it be possible to add type hints here?

File: modin/engines/base/io/text/text_file_dispatcher.py [hunk 18]
```diff
@@ -258,3 +259,157 @@ def _read_rows(
...
+            header_size += max(header) + 1
+
+        return header_size
+
+    @classmethod
+    def _define_usecols(
+        cls,
+        filepath_or_buffer: FilePathOrBuffer,
+        **kwargs,
+    ):
```
Comment (williamma12) [inline modin/engines/base/io/text/text_file_dispatcher.py:313]: Could we add a type hint for the return type here as well?

File: modin/engines/base/io/text/text_file_dispatcher.py [hunk 19]
```diff
@@ -258,3 +259,157 @@ def _read_rows(
...
+
+        return usecols
+
+    @classmethod
+    def _define_metadata(
+        cls,
+        df: pandas.DataFrame,
+        num_splits: int,
+        column_names: Sequence,
+    ):
```
Comment (williamma12) [inline modin/engines/base/io/text/text_file_dispatcher.py:345]: Can we add a return type hint here as well?

**Review by williamma12 — COMMENTED** [review]
This is a really great PR! `CSVDispatcher` is significantly more readable and easier to work with (Though uncle bob from clean code may still scoff at the 120 lines in one function ;) )! I also really like how you moved certain functions to the `TextFileDispatcher` that are helpful in other dispatchers (e.g., `JSONDispatcher` and `FwfDispatcher`). I left some comments mainly regarding type hinting.

This is more for my own understanding but does `_md` mean modified? such as `filepath_or_buffer_md` and `index_col_md`?

File: modin/engines/base/io/text/text_file_dispatcher.py (hunk 9 above)
Comment (amyskov) [reply modin/engines/base/io/text/text_file_dispatcher.py]: Yes, i think, it would, regular `if else` statements were used.

File: modin/engines/base/io/text/text_file_dispatcher.py (hunk 11 above)
Comment (amyskov) [reply modin/engines/base/io/text/text_file_dispatcher.py]: Yes, it could help, but `_validate_usecols_arg` was removed in 479ca1972e75869c4b57f5c64441b12717307227, so there is no need in this function.

File: modin/engines/base/io/text/text_file_dispatcher.py (hunk 13 above)
Comment (amyskov) [reply modin/engines/base/io/text/text_file_dispatcher.py:282]: Documentation mentions fractionally `header=None` case in the string ` if column names are passed explicitly then the behavior is identical to header=None`.
`heder=None` can be set, for example, to ignore inferring of header row:
```
import pandas
from io import StringIO

data = "1,2,3\n4,5,6\n"

df = pandas.read_csv(StringIO(data), header=None)
print(df)

Output:
   0  1  2
0  1  2  3
1  4  5  6
```

File: modin/engines/base/io/text/csv_dispatcher.py (hunk 14 above)
Comment (amyskov) [reply modin/engines/base/io/text/csv_dispatcher.py]: It can be done for `PandasQueryCompiler`, but we also have to add experimental query compiler, that will cause `ImportError` in environments where experimental dependencies are not installed.

File: modin/engines/base/io/text/csv_dispatcher.py (hunk 15 above)
Comment (amyskov) [reply modin/engines/base/io/text/csv_dispatcher.py]: Yes, type hints are added.

File: modin/engines/base/io/text/csv_dispatcher.py (hunk 17 above)
Comment (amyskov) [reply modin/engines/base/io/text/csv_dispatcher.py]: Type hints are added.

File: modin/engines/base/io/text/text_file_dispatcher.py (hunk 18 above)
Comment (amyskov) [reply modin/engines/base/io/text/text_file_dispatcher.py]: Ok,  return type hint was added.

File: modin/engines/base/io/text/text_file_dispatcher.py (hunk 19 above)
Comment (amyskov) [reply modin/engines/base/io/text/text_file_dispatcher.py]: Return type hint was added.

File: modin/engines/base/io/text/text_file_dispatcher.py (hunk 13 above)
Comment (williamma12) [reply modin/engines/base/io/text/text_file_dispatcher.py:282]: Ah ok my mistake! This looks good then!

File: modin/engines/base/io/text/csv_dispatcher.py (hunk 10 above)
Comment (williamma12) [reply modin/engines/base/io/text/csv_dispatcher.py]: that looks good!

File: modin/engines/base/io/text/text_file_dispatcher.py (hunk 11 above)
Comment (williamma12) [reply modin/engines/base/io/text/text_file_dispatcher.py]: ah ok sounds good!

**Review by williamma12 — APPROVED** [review]
Ah ok that makes sense! This looks good with me
